ironfish/src/wallet/wallet.ts

```diff
@@ -1813,10 +1819,56 @@ export class Wallet {
         return
       }
 
+      this.stopUnlockTimeout()
       this.accountById.clear()
       this.locked = true
     } finally {
       unlock()
     }
   }
+
+  async unlock(passphrase: string, timeout?: number, tx?: IDatabaseTransaction): Promise<void> {
+    const unlock = await this.createTransactionMutex.lock()
+
+    try {
+      const encrypted = await this.walletDb.accountsEncrypted(tx)
+      if (!encrypted) {
+        return
+      }
+
+      for (const [id, account] of this.encryptedAccountById.entries()) {
+        this.accountById.set(id, account.decrypt(passphrase))
+      }
+
+      this.startUnlockTimeout(timeout)
+      this.locked = false
```

Comment:
Might be a bit of an overkill defensive thought but, does this get stuck in a weird state if for some unknown/bug reason one of the decrypts fails, or perhaps even a failure to read the encrypted accounts from disk? 

Thinking about it a little, should the `catch` try to re-lock things up or recover in some manner? Or maybe start the unlock timeout in the finally?

Reply:
I don't think reading encrypted accounts from disk should apply since this is using cached values. But if at least one decryption fails, that's true this could end up in a weird state where some subset of accounts would be decrypted.

I don't think starting the timeout should happen in the `finally` since that will always execute, and the re-lock should be dependent on all decryptions being successful. But I think re-locking in the `catch` is a good idea - that will guard if the first decryptions worked but subsequent ones did not.